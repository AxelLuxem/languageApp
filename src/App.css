#root {
  max-width: 100%;
  margin: 0px;
  padding: 0px;
  text-align: center;
}

.colorWrapper {
  display: grid;
  width: 100vw;
  grid-template-columns: 1fr 1fr;
  gap: 6px;
}

.colorSquare {
  height: 50vh;
  /* width: 50vw; */
}
