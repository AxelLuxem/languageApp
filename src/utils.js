import colors from "./components/colorList";

export const getRandomIndex = (numberColors) => {
  return Math.floor(Math.random() * numberColors);
};

export const getColorSet = () => {
  const colorSet = [];

  for (let i = 0; i < 4; i++) {
    let color;
    let alreadyExists;
    do {
      const index = getRandomIndex(colors.length);
      color = colors[index];

      alreadyExists = colorSet.findIndex(
        (item) => item.english === color.english
      );
    } while (alreadyExists > -1);
    colorSet.push(color);
  }

  return colorSet;
};
