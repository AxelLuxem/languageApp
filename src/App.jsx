import { useState, useEffect } from "react";
import "./App.css";

import colors from "./components/colorList";

const scoreAnimationDuration = 0.5;

const App = () => {
  const [score, setScore] = useState(0);
  const [questionColor, setQuestionColor] = useState();
  const [fourColors, setFourColors] = useState([]);
  const [wrong, setWrong] = useState(false);
  const [correct, setCorrect] = useState(false);
  const [soundOn, setSoundOn] = useState(true);

  const generateColor = () => {
    const colorSet = [];

    for (let i = 0; i < 4; i++) {
      let color;
      let alreadyExists;
      do {
        const index = Math.floor(
          Math.random() * colors.length
        );
        color = colors[index];

        alreadyExists = colorSet.findIndex(
          (item) => item.english === color.english
        );
      } while (alreadyExists > -1);
      colorSet.push(color);
    }

    setFourColors(colorSet);
    const questionColorIndex = Math.floor(
      Math.random() * fourColors.length
    );
    const colorObjectSelected =
      colorSet[questionColorIndex];
    setQuestionColor(colorObjectSelected);
  };

  const submitAnswer = (answer) => {
    if (answer === questionColor.french) {
      document.getElementById(
        "Score"
      ).style.animation = `jump-shake ${scoreAnimationDuration}s`;
      setScore(score + 1);
      setQuestionColor();
      setFourColors([]);
      setCorrect(true);

      setTimeout(() => {
        generateColor();
        setCorrect(false);
        document.getElementById("Score").style.animation =
          null;
      }, scoreAnimationDuration * 1000 + 500);
    } else {
      setQuestionColor();
      setFourColors([]);
      setWrong(true);

      setTimeout(() => {
        generateColor();
        setWrong(false);
      }, 1000);
    }
  };

  const playSound = () => {
    const english = questionColor?.english;
    console.log("in the playSounds");

    if (english !== undefined) {
      questionColor.audio.play();
    }
  };

  const keypress = (event) => {
    const key = event.key;
    if (
      key === "1" ||
      key === "2" ||
      key === "3" ||
      key === "4"
    ) {
      const keypressInput =
        document.getElementById("keypressInput");
      keypressInput.value = key;
      keypressInput.focus();
    }
  };

  const createKeyPressListener = () => {
    document.addEventListener("keypress", keypress);
  };

  useEffect(() => {
    if (soundOn) {
      playSound();
    }
  }, [questionColor, soundOn]);

  useEffect(() => {
    generateColor();
    createKeyPressListener();
  }, []);

  return (
    <div className="pageWrapper">
      <div className="heading">
        <div className="block">
          <div className="toggleSwitchWrapper">
            <label className="switchText">Sound</label>
            <label className="switch">
              <input
                id="switch"
                defaultChecked={true}
                onChange={() => setSoundOn(!soundOn)}
                type="checkbox"
              />
              <span className="slider round"></span>
            </label>
          </div>
        </div>
        <div className="scoreWrapper block">
          <p id="Score" className="scoreText">
            Score: {` ${score}`}
          </p>
        </div>
        <div className="block"></div>
      </div>
      <div className="squaresWrapper">
        <div className="test">
          <h2>
            Please click on
            {` ${questionColor?.french || "..."}`}
          </h2>
          <div className="colorGrid">
            {correct ? (
              <img className="answer" src="Checkmark.png" />
            ) : wrong ? (
              <img className="answer" src="Red-X.png" />
            ) : (
              fourColors.map((item, index) => {
                const { english } = item;
                return (
                  <button
                    key={index}
                    style={{
                      backgroundColor: `${english.toLowerCase()}`,
                    }}
                    className="colorSquare"
                    onClick={() =>
                      submitAnswer(item.french)
                    }
                  >
                    {index + 1}
                  </button>
                );
              })
            )}
          </div>
        </div>
      </div>
      <form>
        <input
          id="keypressInput"
          type="text"
          className="hiddenInput"
          onFocus={() => {
            const fourColorsInput =
              document.getElementById("keypressInput");
            const value = fourColorsInput.value;
            submitAnswer(fourColors[value - 1].french);
            document.activeElement.blur();
          }}
        />
      </form>
      <div className="instructionsWrapper">
        <h3>----How to play----</h3>
        <p>
          To play you have to click on colored square that
          corresponds to the color that is written above the
          four squares.
          <br />
          You can either left click on the chosen square, or
          use numbers 1-4 to choose a square
        </p>
        <p>
          If you wish to turn the sound off, then you can do
          so by using the slider at the top left of the page
        </p>
      </div>
    </div>
  );
};

export default App;
