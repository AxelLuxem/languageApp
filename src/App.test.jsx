import { expect, describe, it, vi } from "vitest";
import { render } from "@testing-library/react";
import App from "./App";

import colors from "./components/colorList";
import { getRandomIndex, getColorSet } from "./utils";

global.Audio = vi.fn().mockImplementation(() => ({
  pause: vi.fn(),
  play: vi.fn(() => Promise.resolve()),
}));

vi.mock("./utils");

describe("<App />", () => {
  it("should render the App component", () => {
    vi.mocked(getRandomIndex).mockReturnValue(1);
    vi.mocked(getColorSet).mockReturnValue([
      colors[0],
      colors[1],
      colors[2],
      colors[3],
    ]);

    const app = render(<App />);

    expect(app).toMatchSnapshot();
  });
});
