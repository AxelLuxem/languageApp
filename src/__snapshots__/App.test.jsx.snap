// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<App /> > should render the App component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pageWrapper"
      >
        <div
          class="heading"
        >
          <div
            class="block"
          >
            <div
              class="toggleSwitchWrapper"
            >
              <label
                class="switchText"
              >
                Sound
              </label>
              <label
                class="switch"
              >
                <input
                  checked=""
                  id="switch"
                  type="checkbox"
                />
                <span
                  class="slider round"
                />
              </label>
            </div>
          </div>
          <div
            class="scoreWrapper block"
          >
            <p
              class="scoreText"
              id="Score"
            >
              Score: 
               0
            </p>
          </div>
          <div
            class="block"
          />
        </div>
        <div
          class="squaresWrapper"
        >
          <div
            class="test"
          >
            <h2>
              Please click on
               Vert
            </h2>
            <div
              class="colorGrid"
            >
              <button
                class="colorSquare"
                style="background-color: red;"
              >
                1
              </button>
              <button
                class="colorSquare"
                style="background-color: green;"
              >
                2
              </button>
              <button
                class="colorSquare"
                style="background-color: blue;"
              >
                3
              </button>
              <button
                class="colorSquare"
                style="background-color: yellow;"
              >
                4
              </button>
            </div>
          </div>
        </div>
        <form>
          <input
            class="hiddenInput"
            id="keypressInput"
            type="text"
          />
        </form>
        <div
          class="instructionsWrapper"
        >
          <h3>
            ----How to play----
          </h3>
          <p>
            To play you have to click on colored square that corresponds to the color that is written above the four squares.
            <br />
            You can either left click on the chosen square, or use numbers 1-4 to choose a square
          </p>
          <p>
            If you wish to turn the sound off, then you can do so by using the slider at the top left of the page
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pageWrapper"
    >
      <div
        class="heading"
      >
        <div
          class="block"
        >
          <div
            class="toggleSwitchWrapper"
          >
            <label
              class="switchText"
            >
              Sound
            </label>
            <label
              class="switch"
            >
              <input
                checked=""
                id="switch"
                type="checkbox"
              />
              <span
                class="slider round"
              />
            </label>
          </div>
        </div>
        <div
          class="scoreWrapper block"
        >
          <p
            class="scoreText"
            id="Score"
          >
            Score: 
             0
          </p>
        </div>
        <div
          class="block"
        />
      </div>
      <div
        class="squaresWrapper"
      >
        <div
          class="test"
        >
          <h2>
            Please click on
             Vert
          </h2>
          <div
            class="colorGrid"
          >
            <button
              class="colorSquare"
              style="background-color: red;"
            >
              1
            </button>
            <button
              class="colorSquare"
              style="background-color: green;"
            >
              2
            </button>
            <button
              class="colorSquare"
              style="background-color: blue;"
            >
              3
            </button>
            <button
              class="colorSquare"
              style="background-color: yellow;"
            >
              4
            </button>
          </div>
        </div>
      </div>
      <form>
        <input
          class="hiddenInput"
          id="keypressInput"
          type="text"
        />
      </form>
      <div
        class="instructionsWrapper"
      >
        <h3>
          ----How to play----
        </h3>
        <p>
          To play you have to click on colored square that corresponds to the color that is written above the four squares.
          <br />
          You can either left click on the chosen square, or use numbers 1-4 to choose a square
        </p>
        <p>
          If you wish to turn the sound off, then you can do so by using the slider at the top left of the page
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
